### PR TITLE
ocamlPackages.owl: init at 0.7.1

### PR DESCRIPTION
**Pull request by bcdarwin**
###### Motivation for this change

Add a package and its dependencies (eigen, owl-base).

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [x] other Linux distributions
- [NA] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [NA] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [NA] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

### CODE REVIEW

File: pkgs/development/ocaml-modules/eigen/default.nix [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    repo   = pname;
+    rev    = version;
+    sha256 = "1bn4sgnrdq065xnbkzjh2fpkbnix1h2nnzb2qjbz67pcbls17c1n";
+  };
+
+  propagatedBuildInputs = [ ctypes ];
+
+  meta = with stdenv.lib; {
+    description = "Minimal/incomplete Ocaml interface to Eigen3, mostly for Owl";
+    homepage = "https://github.com/${owner}/${pname}";
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/eigen/default.nix]:
```suggestion
    inherit (src.meta) homepage;
```

File: pkgs/development/ocaml-modules/owl-base/default.nix [hunk 2]
```diff
@@ -0,0 +1,26 @@
+{ stdenv, buildDunePackage, fetchFromGitHub, integers, stdlib-shims }:
+
+buildDunePackage rec {
+  pname = "owl-base";
+  version = "0.7.1";
+
+  owner = "owlbarn";
+
+  src = fetchFromGitHub {
+    inherit owner;
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/owl-base/default.nix]:
```suggestion
    owner = "owlbarn";
```

**Comment by vbgl** [comment]
@GrahamcOfBorg build ocamlPackages.owl

**Comment by vbgl** [comment]
Maybe you could tweak the `EIGEN_FLAGS` environment variable on Aarch64 (see: <https://github.com/owlbarn/eigen/blob/068a74666322f575e32795ee9290b92a30e4a123/eigen/dune#L4-L6>). Currently build fails with:

> gcc: error: unrecognized command line option '-mfpmath=sse'

File: pkgs/development/ocaml-modules/owl/default.nix [hunk 3]
```diff
@@ -0,0 +1,15 @@
+{ stdenv, buildDunePackage, fetchFromGitHub, alcotest
+, eigen, stdio, stdlib-shims, openblasCompat, owl-base
+}:
+
+buildDunePackage rec {
+  pname = "owl";
+
+  inherit (owl-base) version owner src meta minimumSupportedOcamlVersion;
```
[the remark this answers is not in the document]
Comment (bcdarwin) [reply pkgs/development/ocaml-modules/owl/default.nix]: Done.

File: pkgs/development/ocaml-modules/owl-base/default.nix [hunk 4]
```diff
@@ -0,0 +1,26 @@
+{ stdenv, buildDunePackage, fetchFromGitHub, integers, stdlib-shims }:
+
+buildDunePackage rec {
+  pname = "owl-base";
+  version = "0.7.1";
+
+  owner = "owlbarn";
```
[the remark this answers is not in the document]
Comment (bcdarwin) [reply pkgs/development/ocaml-modules/owl-base/default.nix]: Done.

File: pkgs/development/ocaml-modules/owl-base/default.nix (hunk 2 above)
Comment (bcdarwin) [reply pkgs/development/ocaml-modules/owl-base/default.nix]: Done.

File: pkgs/development/ocaml-modules/owl-base/default.nix [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+  src = fetchFromGitHub {
+    inherit owner;
+    repo   = "owl";
+    rev    = version;
+    sha256 = "1v4jfn3w18zq188f9gskx9ffja3xx59j2mgrw6azp8lsbqixg5xk";
+  };
+
+  propagatedBuildInputs = [ stdlib-shims ];
+
+  minimumSupportedOcamlVersion = "4.06.0";
```
[the remark this answers is not in the document]
Comment (bcdarwin) [reply pkgs/development/ocaml-modules/owl-base/default.nix]: OK.

[the remark this answers is not in the document]
Comment (vbgl) [reply pkgs/development/ocaml-modules/owl-base/default.nix]: OCaml is spelled with a capital C.

[the remark this answers is not in the document]
Comment (bcdarwin) [reply pkgs/development/ocaml-modules/owl-base/default.nix]: Fixed.

**Comment by bcdarwin** [comment]
@vbgl -- flags adjusted on Aarch64.

**Comment by bcdarwin** [comment]
Owl itself has the same issue with SSE-related flags, so it might make more sense to set `platforms` to `x86_64` rather than make all these changes manually.

**Comment by vbgl** [comment]
That’s up to you, as maintainer of these packages: do you want to figure out the right installation parameters for these architectures that you don’t use, or just claim that the package is only available for the architecture that you use.

Minor notes:

 - [eigen-0.1.5](https://github.com/owlbarn/eigen/releases/tag/0.1.5) is out
 - You don’t have to “substituteInPlace” to change the compilation flags: simply setting the “EIGEN_FLAGS” and “OPTFLAGS0” environment variables should be enough.

**Comment by bcdarwin** [comment]
Since I have no way to test on these architectures, I've just set platforms to x64 for now, which together with the Eigen version bump simplifies the derivation considerably.

**Comment by vbgl** [comment]
You’ve made a good choice.

Final nitpick: eigen won’t build with OCaml < 4.04 Can you please add a “minimumOCamlVersion” to that derivation? Thanks.

**Comment by bcdarwin** [comment]
@vbgl - done!

**Comment by vbgl** [comment]
Thanks!

**Pull request merged**